cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetectorState.java

```diff
@@ -80,13 +98,16 @@ void addAnomalyDetection(AnomalyType anomalyType, Anomaly anomaly) {
    * @param anomaly The anomaly to handle.
    * @param status A status information regarding how the anomaly was handled.
    */
-  void onAnomalyHandle(Anomaly anomaly, AnomalyState.Status status) {
+  synchronized void onAnomalyHandle(Anomaly anomaly, AnomalyState.Status status) {
     AnomalyType anomalyType = getAnomalyType(anomaly);
     String anomalyId = anomaly.anomalyId();
 
-    AnomalyState recentAnomalyState = _recentAnomaliesByType.get(anomalyType).get(anomalyId);
+    AnomalyState recentAnomalyState = _recentAnomaliesByType.get(anomalyType).remove(anomalyId);
     if (recentAnomalyState != null) {
       recentAnomalyState.setStatus(status);
+      if (status == AnomalyState.Status.FIX_STARTED) {
+        _ongoingSelfHealingAnomalyState = recentAnomalyState;
```

Comment:
looks like we [add the anomaly to _recentAnomaliesByType](_recentAnomaliesByType) and we [remove it from _recentAnomaliesByType and mark it as ongoing anomaly](https://github.com/linkedin/cruise-control/blob/b0c81b29e15a6006211da6d79a17846ab1395aad/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetector.java#L280) are in the same method and it is executed in a [dedicated thread](https://github.com/linkedin/cruise-control/blob/b0c81b29e15a6006211da6d79a17846ab1395aad/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetector.java#L158), I feel in this way as long as the `LinkedHashMap` used in  `_recentAnomaliesByType` has size larger than 1, we will not get null from this [remove](https://github.com/linkedin/cruise-control/blob/b0c81b29e15a6006211da6d79a17846ab1395aad/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetectorState.java#L105).

However, I think the suggestion of switching to use  ongoing anomaly has a benefit of making response more readable. If we return  `ongoingSelfHealingAnomalyState` in `anomalyState` response, we probably return the same thing in `recentAnomaly` part, so we return the same thing twice and it could be very long since for `GoalViolation` and `BrokerFailure` it has the optimization result.

Reply:
1. I agree, it sounds reasonable to assume that `addAnomalyDetection()` and `onAnomalyHandle(...)` will always be handled by a single thread.
2. Switching to use ongoing anomaly sounds good to me!

One final question: Why do we remove the anomaly state from `_recentAnomaliesByType` rather than simply getting its value? -- i.e. why `_recentAnomaliesByType.get(anomalyType).remove(anomalyId);` but not `_recentAnomaliesByType.get(anomalyType).get(anomalyId);`?

I suspect that removing it from the recent anomalies will lead to loss of anomaly state upon completion of the ongoing execution (i.e. anomaly fix)?

Reply:
Ah, good catch. Previously since we want to keep ongoing anomaly state in `_ongoingSelfHealingAnomalyState`, it make sense to remove it from `_recentAnomaliesByType`, now we do not need to remove since we switch to keep ongoing anomaly.